A55-stateful-session-affinity.md

```diff
@@ -0,0 +1,398 @@
+sample configuration (expressed as yaml) is as follows:
+
+```yaml
+   http_filters:
+   - name: envoy.filters.http.stateful_session
+     typed_config:
+       "@type": type.googleapis.com/envoy.extensions.filters.http.stateful_session.v3.StatefulSession
+       session_state:
+         name: envoy.http.stateful_session.cookie
+         typed_config:
+           "@type": type.googleapis.com/envoy.extensions.http.stateful_session.cookie.v3.CookieBasedSessionState
+           cookie:
+             name: global-session-cookie
+             path: /Package1.Service2/Method3
+             ttl: 120s
+```
+
+`name` is the name of the cookie used in this feature.
+
+`path` is the path for which this cookie is applied. **Note:** the path value
+here needs to match any path specified in a RouteConfiguration when route
+based [Filter Config Override][filter_config_override] is used.
+
+
+`ttl` is the time-to-live for the cookie. It is set on the cookie but will not
+be enforced by gRPC.
+
+This new filter will be added to the HTTP Filter Registry and processed as
+described in [A39: xDS HTTP Filter Support][A39]. Specifically when this filter
+is present in the configuration, gRPC will create the appropriate client filter
+(aka interceptor in Java/Go) and install it in the channel to process data
+plane RPCs. We call this filter `CookieBasedStatefulSessionFilter`. It will
+copy the 3 configuration values - `name`, `ttl`  and `path` -  into the filter
+object. Validation and processing are further described in
+[Stateful session][stateful-session].
+
+Note that
+[StatefulSessionPerRoute][stateful-session-per-route] *will* be supported as
+[Filter Config Overrides][filter_config_override]. For example, let's say
+the top level config (in `HttpConnectionManager.http_filters`) contains certain
+`StatefulSession` configuration. This configuration can either be
+overridden or disabled by a
+[StatefulSessionPerRoute][stateful-session-per-route] config for a particular
+virtual host or route through a merged configuration as described in
+[A39 Filter Config Overrides][filter_config_override].
+
+The filter is described in more detail in the section
+[`CookieBasedStatefulSessionFilter` Processing][filter-section].
+
+### Load Balancer Configuration Containing `override_host_status`
+
+Even with stateful session affinity enabled, gRPC will only send an RPC to a
+valid backend (host) as defined by
+[`common_lb_config.override_host_status`][or-host-status]. When the backend
+state is invalid, gRPC will use the configured load balancing policy to pick
+the backend.
+
+Note that gRPC currently ignores endpoints with health status other than
+`HEALTHY` or `UNKNOWN` and even `XdsClient` excludes such endpoints in its
+update to the watchers. We will need a separate design to include such
+endpoints if we determine that such behavior is needed for this feature to
+be useful to our users.
+
+As a result, for this design gRPC will only consider the `UNKNOWN` or
```

Comment:
Have we considered what happens if `override_host_status` includes `UNKNOWN` but not `HEALTHY`, or vice-versa?  We don't currently plumb the health state of the endpoints out of the `XdsClient`, so if we need to differentiate between these two states, we'd need to do that somehow.

I think for this design, we should just say that if either of those two states is present in the set, then we treat it as if both states are present.  In other words, this will act as a simple boolean toggle: if either `UNKNOWN` or `HEALTHY` is present in the set, then we enable the host-override functionality; otherwise, we disable it.

Reply:
what about the default cause? when the value of override_host_status is not set it should be equivalent to  [UNKNOWN, HEALTHY, DEGRADED]. In other words it should be disabled only if override_host_status is set with an an empty set.

Reply:
If the field is unset, we treat it the same as if it was explicitly set to the default set.

Under this current design approach, the feature will be disabled whenever the set does not include either `HEALTHY` or `UNKNOWN`.  This can happen when the set is empty, but it could also happen if the set contains only `UNHEALTHY`.

Reply:
> If the field is unset, we treat it the same as if it was explicitly set to the default set.

So even when the field is unset, we will incur the overhead of this enabled policy. Just confirming.